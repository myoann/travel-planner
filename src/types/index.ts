type TLanguage = {
  code: string;
  name: string;
  native: string;
  rtl: boolean;
};

type TSubdivision = {
  code: string;
  emoji: string;
  name: string;
};

type TState = {
  code: string;
};

type TContinent = {
  code: string;
  countries: {
    code: string;
  }[];
  name: string;
};

export type TCountry = {
  awsRegion: string;
  capital: string;
  code: string;
  continent: TContinent;
  currencies: string[];
  currency: string;
  emoji: string;
  emojiU: string;
  languages: TLanguage[];
  name: string;
  native: string;
  phone: string;
  phones: string[];
  states: TState[];
  subdivisions: TSubdivision[];
};

export type TOption = {
  key: string;
  name: string;
  value: string;
};
