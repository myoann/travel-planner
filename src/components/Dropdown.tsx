"use client";

import { useState } from "react";
import Image from "next/image";

import { TOption } from "@/types";

type TProps = {
  defaultSelectName: string;
  onClick: (value: TOption) => void;
  options: TOption[];
  selectedOption?: TOption | null;
};

export default function Dropdown({
  defaultSelectName,
  onClick,
  options,
  selectedOption,
}: TProps) {
  const [isOpen, setIsOpen] = useState(false);

  /** Toggles the dropdown visibility */
  const toggleDropdown = () => {
    setIsOpen(!isOpen);
  };

  /** Handles the click on an option */
  const onClickOption = (value: TOption) => {
    onClick(value);
    setIsOpen(false);
  };

  return (
    <div className="relative inline-block text-left w-64 z-30">
      <button
        className="inline-flex w-full justify-center gap-x-1.5 px-3 py-2 text-sm font-semibold text-gray-900 hover:text-gray-950 hover:font-bold"
        id="menu-button"
        data-testid="dropdown"
        aria-expanded="true"
        aria-haspopup="true"
        onClick={toggleDropdown}
      >
        {selectedOption?.name || defaultSelectName}

        <Image src="arrowDown.svg" alt="Arrow down" width={20} height={20} />
      </button>

      <div
        className={`${
          isOpen ? "block" : "hidden"
        } absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none py-1`}
        role="menu"
        aria-orientation="vertical"
        aria-labelledby="menu-button"
        tabIndex={-1}
      >
        {options.map((option) => (
          <button
            className={`w-full text-left block px-4 py-2 text-sm hover:bg-gray-100 hover:text-gray-900 text-gray-700 ${
              selectedOption?.value === option.value &&
              "bg-gray-200 text-gray-900 font-bold hover:bg-gray-200"
            }`}
            role="menuitem"
            tabIndex={-1}
            id={`menu-item-${option.key}`}
            data-testid="dropdown-option"
            key={option.key}
            onClick={() => onClickOption(option)}
          >
            {option.name}
          </button>
        ))}
      </div>
    </div>
  );
}
