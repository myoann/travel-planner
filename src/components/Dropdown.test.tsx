import { render, screen } from "@testing-library/react";

import { TOption } from "@/types";

import Dropdown from "./Dropdown";

const defaultProps = {
  defaultSelectName: "Filter",
  onClick: jest.fn(),
  options: [
    {
      key: "1",
      name: "First option",
      value: "first",
    },
  ],
  selectedOption: null,
};

describe("Dropdown", () => {
  it("matches the snapshot", () => {
    const { container } = render(<Dropdown {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it("has a button which contains the text from the props defaultSelectName if no selectedOption is given", () => {
    render(<Dropdown {...defaultProps} />);

    const button = screen.getByRole("button");
    expect(button).toHaveTextContent(defaultProps.defaultSelectName);
  });

  it("has a button which contains the text from the props selectedOption when it is given", () => {
    const selectedOption = defaultProps.options[0];
    render(<Dropdown {...defaultProps} selectedOption={selectedOption} />);

    const button = screen.getByRole("button");
    expect(button).toHaveTextContent(selectedOption.name);
  });

  it("has as many buttons as there are options in the props", async () => {
    const options: TOption[] = [
      {
        key: "1",
        name: "First option",
        value: "first",
      },
      {
        key: "2",
        name: "Second option",
        value: "second",
      },
      {
        key: "3",
        name: "Third option",
        value: "third",
      },
    ];
    render(<Dropdown {...defaultProps} options={options} />);

    const buttons = screen.getAllByTestId("dropdown-option");
    expect(buttons).toHaveLength(options.length);
  });
});
