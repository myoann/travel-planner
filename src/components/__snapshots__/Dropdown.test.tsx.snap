// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches the snapshot 1`] = `
<div>
  <div
    class="relative inline-block text-left w-64 z-30"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      class="inline-flex w-full justify-center gap-x-1.5 px-3 py-2 text-sm font-semibold text-gray-900 hover:text-gray-950 hover:font-bold"
      data-testid="dropdown"
      id="menu-button"
    >
      Filter
      <img
        alt="Arrow down"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="arrowDown.svg"
        style="color: transparent;"
        width="20"
      />
    </button>
    <div
      aria-labelledby="menu-button"
      aria-orientation="vertical"
      class="hidden absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none py-1"
      role="menu"
      tabindex="-1"
    >
      <button
        class="w-full text-left block px-4 py-2 text-sm hover:bg-gray-100 hover:text-gray-900 text-gray-700 false"
        data-testid="dropdown-option"
        id="menu-item-1"
        role="menuitem"
        tabindex="-1"
      >
        First option
      </button>
    </div>
  </div>
</div>
`;
